### PR TITLE
NavigationViewController Responsiveness

### PR DESCRIPTION
**Pull request by JThramer**
This implements part of #830, where I'm creating a "Nav-bar" demo in [navigation-ios-examples](https://github.com/mapbox/navigation-ios-examples) where the NavigationViewController is pushed onto a `UINavigationController` VC stack. So far, there are three main issues that I see surrounding responsiveness:

- [x] NavigationViewController is not subclassable outside of the `MapboxNavigation` module.
- [x] NavigationViewController should be responsive to being pushed as opposed to presented ("X" button)
- [x] ~NavigationViewController should be able to be instantiated in Interface Builder.~ (punted to #968)

### CODE REVIEW

**Comment by JThramer** [comment]
So after talking with @frederoni it sounds like the plan here is to spin off that third task into it's own ticket (it's not critical to this story), and get this PR in as-is so we can focus on the "embedded view" part of #830.

File: MapboxNavigation/NavigationViewController.swift [hunk 1]
```diff
@@ -154,6 +154,19 @@ public protocol NavigationViewControllerDelegate {
...
+     Returns whether the `BottomBannerView` in the `RouteMapViewController` should show a cancel button.
+     
+     If implemented, this method is called at `RouteMapViewController`'s `viewWillAppear()` time. This is used to initally determine whether the cancel button should be shown when the view first appears. If you need to set the visibility of the cancel button after this point, manipulate the `BottomBannerView.isShowingCancelButton` property directly.
+     
+     - parameter navigationViewController: The navigation view controller that needs to know wheather or not to show the cancel button.
+     - parameter bottomBanner: The bottom banner containing the cancel button.
+     - note: If this method is not implemented, the default behavior is to show the cancel button if the `NavigationViewController` was modally presented.
+     - returns: True to show the cancel button, False to hide.
+     */
+    @objc optional func navigationViewController(_ navigationViewController: NavigationViewController, shouldShowCancelButtonIn bottomBanner: BottomBannerView) -> Bool
```
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift]: This method bridges to Objective-C as `-navigationViewController:shouldShowCancelButtonIn:`. To conform to Objective-C naming conventions, we’ll need `@objc(navigationViewController:shouldShowCancelButtonInBottomBannerView:)`.

However, I’m not sure about passing the BottomBannerView as a parameter. Can you think of a situation in which the implementation would need to consult a property on BottomBannerView to decide whether to return true or false? If that does indeed turn out to be useful, we should make the BottomBannerView accessible via NavigationViewController, ideally by making `NavigationViewController.routeMapViewController` public. That would be useful in its own right.

File: MapboxNavigation/NavigationViewController.swift [hunk 2]
```diff
@@ -420,6 +433,7 @@ open class NavigationViewController: UIViewController, RouteMapViewControllerDel
         UIApplication.shared.applicationIconBadgeNumber = 0
     }
     
+    //MARK: - RouteMapViewControllerDelegate Methods
```
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift]: A common practice in this codebase is to break out a separate extension and have it conform to the protocol. That way, the compiler will complain if any of the protocol’s required methods lies outside the extension.

File: MapboxNavigation/RouteMapViewController.swift [hunk 3]
```diff
@@ -961,6 +964,7 @@ extension RouteMapViewController: StatusViewDelegate {
     }
 }
 
+//MARK: - RouteMapViewControllerDelegate
```
Comment (1ec5) [inline MapboxNavigation/RouteMapViewController.swift]: This protocol already gets its own section in the jump bar menu. A documentation comment would be useful, however. 😉

File: MapboxNavigation/RouteMapViewController.swift [hunk 4]
```diff
@@ -976,9 +980,19 @@ protocol RouteMapViewControllerDelegate: class {
...
     func mapViewControllerDidCancelNavigation(_ mapViewController: RouteMapViewController)
+    func mapViewController(_ mapViewController: RouteMapViewController, shouldShowCancelButtonIn bottomBanner: BottomBannerView) -> Bool
     func mapViewController(_ mapViewController: RouteMapViewController, didSend feedbackId: String, feedbackType: FeedbackType)
     
     func mapViewController(_ mapViewController: RouteMapViewController, mapViewUserAnchorPoint mapView: NavigationMapView) -> CGPoint?
     
     func mapViewControllerShouldAnnotateSpokenInstructions(_ routeMapViewController: RouteMapViewController) -> Bool
 }
+
+extension RouteMapViewControllerDelegate where Self: UIViewController {
```
Comment (1ec5) [inline MapboxNavigation/RouteMapViewController.swift]: Note that Objective-C doesn’t support generics on protocols, nor default method implementations in protocols. These limitations may become relevant if we decide to make RouteMapViewController public.

File: MapboxNavigation/NavigationViewController.swift [hunk 5]
```diff
@@ -154,6 +154,19 @@ public protocol NavigationViewControllerDelegate {
      Returns the center point of the user course view in screen coordinates relative to the map view.
      */
     @objc optional func navigationViewController(_ navigationViewController: NavigationViewController, mapViewUserAnchorPoint mapView: NavigationMapView) -> CGPoint
+    
+    /**
+     Returns whether the `BottomBannerView` in the `RouteMapViewController` should show a cancel button.
+     
+     If implemented, this method is called at `RouteMapViewController`'s `viewWillAppear()` time. This is used to initally determine whether the cancel button should be shown when the view first appears. If you need to set the visibility of the cancel button after this point, manipulate the `BottomBannerView.isShowingCancelButton` property directly.
```
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift]: Is an explicit setup method necessary? This note makes it sound like the developer can accomplish the same effect (from the user’s perspective) by setting `isShowingCancelButton` between the call to `init(for:directions:styles:locationManager:)` and `viewDidAppear()`, such as in `prepare(for:)`.

**Comment by JThramer** [comment]
This is ultra-stale, so I'm going to close. If this was a mistake, feel free to re-open.

**Pull request closed without merging**